modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3Repository.java

```diff
@@ -459,4 +474,83 @@ public String getAnalysisFailureExtraDetail() {
             ReferenceDocs.S3_COMPATIBLE_REPOSITORIES
         );
     }
+
+    // only one multipart cleanup process running at once
+    private final AtomicBoolean multipartCleanupInProgress = new AtomicBoolean();
+
+    @Override
+    public void deleteSnapshots(
+        Collection<SnapshotId> snapshotIds,
+        long repositoryDataGeneration,
+        IndexVersion minimumNodeVersion,
+        SnapshotDeleteListener snapshotDeleteListener
+    ) {
+        getMultipartUploadCleanupListener(
+            isReadOnly() ? 0 : MAX_MULTIPART_UPLOAD_CLEANUP_SIZE.get(getMetadata().settings()),
+            new ActionListener<>() {
+                @Override
+                public void onResponse(ActionListener<Void> multipartUploadCleanupListener) {
+                    S3Repository.super.deleteSnapshots(
+                        snapshotIds,
+                        repositoryDataGeneration,
+                        minimumNodeVersion,
+                        new SnapshotDeleteListener() {
+                            @Override
+                            public void onDone() {
+                                snapshotDeleteListener.onDone();
+                            }
+
+                            @Override
+                            public void onRepositoryDataWritten(RepositoryData repositoryData) {
+                                multipartUploadCleanupListener.onResponse(null);
+                                snapshotDeleteListener.onRepositoryDataWritten(repositoryData);
+                            }
+
+                            @Override
+                            public void onFailure(Exception e) {
+                                multipartUploadCleanupListener.onFailure(e);
+                                snapshotDeleteListener.onFailure(e);
+                            }
+                        }
+                    );
+                }
+
+                @Override
+                public void onFailure(Exception e) {
+                    logger.warn("failed to get multipart uploads for cleanup during snapshot delete", e);
+                    snapshotDeleteListener.onFailure(e);
```

Comment:
Does this mean that failures happening while getting list of multipart uploads would also fail the snapshot deletion? `getMultipartUploadCleanupListener` doesn't seem to throw. Sorry if this is not related, `ActionListener<ActionListener<>>` is really bumping this up to a whole new level! :)))

Reply:
> Does this mean that failures happening while getting list of multipart uploads would also fail the snapshot deletion?

Hm that was how it worked in an earlier iteration but in the finished version we log a failure there but otherwise treat it as if there are no multipart uploads. Let me try and simplify that, sec.

> `ActionListener<ActionListener<>>` is really bumping this up to a whole new level

<img width="487" alt="image" src="https://github.com/user-attachments/assets/7bdd9679-daf7-426c-8f3f-7718435b56f4">

Reply:
> Let me try and simplify that, sec.

Actually this is a little tricky, in principle we could turn this into a `Consumer<ActionListener<Void>>` but then we submit a task to a threadpool executor and that sort of thing can theoretically fail. It won't actually fail today because `snapshotExecutor` has an infinite queue and just silently drops work when shut down rather than rejecting it, but still I would prefer to write code like this in a style that allows for failures.

Reply:
I'm ok with the `ActionListener<ActionListener<>>` part. FWIW, probably returning a specialized object with more clear names might make it more readable in my opinion, but that would anyway underneath be an action listener. But why log and propagate the failure if it's not supposed to happen?

Reply:
If we're wrong (e.g. some future change to threadpool behaviour makes us wrong) then we have to do _something_ in production when assertions are disabled, this seems like the most sensible thing to do. There's loads of other spots where we `assert false` to indicate to the reader that we think something is impossible, but nonetheless react  in a somewhat reasonable fashion when assertions are disabled.